pinot-spi/src/test/java/org/apache/pinot/spi/env/SegmentMetadataPropertyConfigTest.java

```diff
@@ -0,0 +1,213 @@
+        CommonsConfigurationUtils.getSegmentMetadataFromFile(CONFIG_FILE, true, true);
+
+    // setting the random value of the test keys
+    for (String key: TEST_PROPERTY_KEY) {
+      configuration.setProperty(key, RandomStringUtils.randomAscii(5));
+    }
+    // recovered keys from the configuration.
+    List<String> recoveredKeys = CommonsConfigurationUtils.getKeys(configuration);
+    testPropertyKeys(recoveredKeys, TEST_PROPERTY_KEY);
+
+    CommonsConfigurationUtils.saveSegmentMetadataToFile(configuration, CONFIG_FILE, null); // save the configuration.
+
+    // reading the configuration from saved file.
+    configuration = CommonsConfigurationUtils.getSegmentMetadataFromFile(CONFIG_FILE, true, true);
+    recoveredKeys = CommonsConfigurationUtils.getKeys(configuration);
+    testPropertyKeys(recoveredKeys, TEST_PROPERTY_KEY);
+  }
+
+  @Test
+  public void testSegmentMetadataPropertyConfigurationWithHeader()
+      throws ConfigurationException {
+    PropertiesConfiguration configuration =
+        CommonsConfigurationUtils.getSegmentMetadataFromFile(CONFIG_FILE, true, true);
+
+    // setting the random value of the test keys
+    for (String key: TEST_PROPERTY_KEY) {
+      configuration.setProperty(key, RandomStringUtils.randomAscii(5));
+    }
+    // recovered keys from the configuration.
+    List<String> recoveredKeys = CommonsConfigurationUtils.getKeys(configuration);
+    testPropertyKeys(recoveredKeys, TEST_PROPERTY_KEY);
+
+    // save the configuration.
+    CommonsConfigurationUtils.saveSegmentMetadataToFile(configuration, CONFIG_FILE,
+        PropertyIOFactoryKind.SegmentMetadataIOFactory.getVersion());
+
+    // reading the configuration from saved file.
+    configuration = CommonsConfigurationUtils.getSegmentMetadataFromFile(CONFIG_FILE, true, true);
+    recoveredKeys = CommonsConfigurationUtils.getKeys(configuration);
+    // assert that Header is not null for the config.
+    assertNotNull(configuration.getHeader());
+
+    // assert that configuration has SegmentMetadataPropertyIOFactory
+    assertEquals(configuration.getIOFactory().getClass(), SegmentMetadataPropertyIOFactory.class);
+    testPropertyKeys(recoveredKeys, TEST_PROPERTY_KEY);
+  }
+
+  @Test
+  public void testSegmentMetadataReaderWithSpecialCharsPropertyKeys()
+      throws ConfigurationException {
+    PropertiesConfiguration configuration =
+        CommonsConfigurationUtils.getSegmentMetadataFromFile(CONFIG_FILE, true, true);
+
+    // setting the random value of the test keys
+    for (String key: TEST_PROPERTY_KEY_WITH_SPECIAL_CHAR) {
+      configuration.setProperty(key, RandomStringUtils.randomAscii(5));
+    }
+    // recovered keys from the configuration.
+    List<String> recoveredKeys = CommonsConfigurationUtils.getKeys(configuration);
+    testPropertyKeys(recoveredKeys, TEST_PROPERTY_KEY_WITH_SPECIAL_CHAR);
+
+    CommonsConfigurationUtils.saveSegmentMetadataToFile(configuration, CONFIG_FILE, null); // save the configuration.
```

Comment:
add a test withHeader?

also since the tests for withHeader and withouHeader are same but there, we can define a dataProvider to parameterize the test methods to save some codes here.